### PR TITLE
Users, Conversations, and Messages now update live between clients

### PR DESCRIPTION
**Pull request by MalikG15**

### CODE REVIEW

File: src/codeu/chat/client/ClientConversation.java [hunk 1]
```diff
@@ -140,13 +142,19 @@ public void startConversation(String title, Uuid owner) {
 	System.out.println(this.getClass().toString() + " startConversation()");
     final boolean validInputs = isValidTitle(title);
 
-    final Conversation conv = (validInputs) ? controller.newConversation(title, owner) : null;
+    final Conversation conv = (validInputs) ? view.newConversation(title, owner) : null;
```
Comment (gbrown3) [inline src/codeu/chat/client/ClientConversation.java:145]: Why is the view's version of the newConversation method being called rather than the controller's version?

File: src/codeu/chat/client/ClientConversation.java [hunk 2]
```diff
@@ -225,22 +234,38 @@ public int conversationsCount() {
...
-      summariesByUuid.put(cs.id, cs);
-      summariesSortedByTitle.insert(cs.title, cs);
-      summariesSortedByCreationTime.put(cs, cs.title);
-    }
-
-    if (currentChanged) {
-      updateCurrentConversation();
-      messageContext.resetCurrent(true);
-    }
+	System.out.println(this.getClass().toString() + " updateAllConversations()");
```
Comment (gbrown3) [inline src/codeu/chat/client/ClientConversation.java:237]: Are you sure you don't want this to be in a log rather than just printed out?

File: src/codeu/chat/client/ClientMessage.java [hunk 3]
```diff
@@ -108,11 +107,12 @@ public void addMessage(Uuid author, Uuid conversation, String body) {
 	System.out.println(this.getClass().toString() + " addMessage()");
     final boolean validInputs = isValidBody(body) && (author != null) && (conversation != null);
 
-    final Message message = (validInputs) ? controller.newMessage(author, conversation, body) : null;
+    final Message message = (validInputs) ? view.newMessage(author, conversation, body) : null;
```
Comment (gbrown3) [inline src/codeu/chat/client/ClientMessage.java:110]: Same question about using the view's version of this method rather than the controller's

File: src/codeu/chat/client/ClientMessage.java [hunk 4]
```diff
@@ -38,9 +39,7 @@
 
   private Message current = null;
 
-  private final Map<Uuid, Message> messageByUuid = new HashMap<>();
-
-  private Conversation conversationHead;
+  public Conversation conversationHead;
```
Comment (gbrown3) [inline src/codeu/chat/client/ClientMessage.java:42]: What does the conversationHead represent again? If there's more information than just the name of the user's conversation stored in it, I would think you wouldn't want it to be public.

File: src/codeu/chat/client/ClientUser.java [hunk 5]
```diff
@@ -35,10 +36,11 @@
 
   private User current = null;
 
-  private final Map<Uuid, User> usersById = new HashMap<>();
+  public final Map<Uuid, User> usersById = new HashMap<>();
```
Comment (gbrown3) [inline src/codeu/chat/client/ClientUser.java:40]: I'm not sure a list of users by their ID should have public access. It seems to me like it should be private, so no one can access user information.

File: src/codeu/chat/client/ClientUser.java [hunk 6]
```diff
@@ -102,7 +108,9 @@ public boolean addUser(String name, String password) {
     boolean validInputs = isValidInput(name);
 
 
-    final User user = (validInputs) ? controller.newUser(name, password) : null;
+    // changed from controller to view
+    final User user = (validInputs) ? view.newUser(name, password) : null;
```
Comment (gbrown3) [inline src/codeu/chat/client/ClientUser.java:114]: Same question about using the view's version of newUser() rather than the controller's version

**Review by gbrown3 — COMMENTED** [review]
I had a few small questions, but otherwise it looks great!

File: src/codeu/chat/ServerMain.java [hunk 7]
```diff
@@ -33,6 +35,8 @@
 final class ServerMain {
 
   private static final Logger.Log LOG = Logger.newLog(ServerMain.class);
+  
+  private static int myPort;
```
Comment (GrayTurtle) [inline src/codeu/chat/ServerMain.java:39]: Do these need to be static class variables? Can this just be in main? I don't think I see it being used anywhere else other than main, but I may be wrong.

File: src/codeu/chat/ServerMain.java [hunk 8]
```diff
@@ -46,7 +50,7 @@ public static void main(String[] args) {
 
     LOG.info("============================= START OF LOG =============================");
 
-    final int myPort = Integer.parseInt(args[2]);
+    myPort = Integer.parseInt(args[2]);
     final byte[] secret = Secret.parse(args[1]);
```
Comment (GrayTurtle) [inline src/codeu/chat/ServerMain.java:54]: What's a secret?

Comment (GrayTurtle) [reply src/codeu/chat/ServerMain.java:54]: Oh wait this is one of the parameters for that secret team code isn't it?

File: src/codeu/chat/client/ClientConversation.java (hunk 2 above)
Comment (GrayTurtle) [reply src/codeu/chat/client/ClientConversation.java:237]: This goes along with all the other method calls that goes out to the command line.

File: src/codeu/chat/client/View.java [hunk 9]
```diff
@@ -40,24 +47,169 @@
...
+  public MainChatPage mainChatPage;
+  
+
+  /**
+   * ClientConnections manages the clients connection on the client side
+   * by checking for an incoming information and parsing it accordingly.
+   * @author malikg
+   *
+   */
+  class ClientConnection extends Thread {
```
Comment (GrayTurtle) [inline src/codeu/chat/client/View.java:92]: [insert Google's favorite question]. So are you gonna work on implementing a better way rather than having a timer? Or at least explain it so we can document an alternative and better way? :D

**Comment by KevinWorkman** [comment]
@GrayTurtle Just a minor note: You probably want to wait until the comments have been resolved before merging the pull request. Malik can still respond here, but it might be more annoying to make changes now.

File: src/codeu/chat/ClientMain.java [hunk 10]
```diff
@@ -44,8 +44,15 @@ public static void main(String [] args) {
...
-    final Controller controller = new Controller(source);
-    final View view = new View(source);
+    Controller controller = null;
+    View view = null;
+    try {
+	    controller = new Controller(source.connect());
+	    view = new View(source.connect());
+    }
+    catch (IOException ex) {
+    	LOG.info("Error connecting to server", ex);
```
Comment (KevinWorkman) [inline src/codeu/chat/ClientMain.java:54]: When would this error realistically occur? Do we want to communicate something to the user?

File: src/codeu/chat/ServerMain.java [hunk 11]
```diff
@@ -93,19 +97,48 @@ private static void runServer(Uuid id,
...
+    		    ConnectionSource serverSourceCopy = null;
+    		    // The given serverSource returns an error when connected
+    		    // so a new one is initiated.
+    			try {
+    				serverSourceCopy = ServerConnectionSource.forPort(myPort);
+    			}
+    			catch (IOException ex) {
+    				 LOG.error(ex, "Failed to create server on port");
+    			}
+        		while (true) {
```
Comment (KevinWorkman) [inline src/codeu/chat/ServerMain.java:119]: I would add a ton more comments to this section. The connection pool is probably one of the more complicated things in the code base. Remember how long it took us to figure it out? Make it easier for the next person by adding comments!

File: src/codeu/chat/SimpleGuiClientMain.java [hunk 12]
```diff
@@ -44,21 +45,25 @@ public static void main(String [] args) {
...
-      LOG.error(ex, "Exception setting up client.");
-    }
+		    try (
+		      final ConnectionSource source = new ClientConnectionSource(address.host, address.port)
+		    ) {
+		      // Creating one single connection for the entire
+		      // client session and sending that to both the controller
+		      // and view.
+		      Connection client = source.connect();
+		      final Controller controller = new Controller(client);
```
Comment (KevinWorkman) [inline src/codeu/chat/SimpleGuiClientMain.java:55]: Again, more comments. What does each of these variables really do? How are they related?

File: src/codeu/chat/client/ClientContext.java [hunk 13]
```diff
@@ -30,6 +30,10 @@ public ClientContext(Controller controller, View view) {
     user = new ClientUser(controller, view);
     conversation = new ClientConversation(controller, view, user);
     message = new ClientMessage(controller, view, user, conversation);
+    // Giving reference to the clientContext
+    // from the view to add information to 
+    // users, conversations, and messages
+    view.clientContext = this;
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientContext.java:36]: Would probably be better to use a setter function or pass in `this` as a constructor parameter.

File: src/codeu/chat/client/ClientConversation.java [hunk 14]
```diff
@@ -41,6 +45,8 @@
 
   private final ClientUser userContext;
   private ClientMessage messageContext = null;
+  
+  private final Map<Uuid, Conversation> conversationsByUuid = new HashMap<>();
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientConversation.java:49]: I'd add JavaDoc explaining exactly what this `Map` is used for. Same for other similar variables.

Comment (KevinWorkman) [reply src/codeu/chat/client/ClientConversation.java:49]: And while I'm being annoying, I'll throw in the idea of scalability into the mix: how large could this `Map` get? What if you had a million users with a million conversations each? How would you change your code to handle that case?

File: src/codeu/chat/client/ClientConversation.java [hunk 15]
```diff
@@ -140,13 +142,19 @@ public void startConversation(String title, Uuid owner) {
...
+    final Conversation conv = (validInputs) ? view.newConversation(title, owner) : null;
 
     if (conv == null) {
       System.out.format("Error: conversation not created - %s.\n",
           (validInputs) ? "server failure" : "bad input value");
+      return;
     } else {
       LOG.info("New conversation: Title= \"%s\" UUID= %s", conv.title, conv.id);
+      
+      summariesByUuid.put(conv.summary.id, conv.summary);
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientConversation.java:154]: Comments on each of these explaining why you need each index would be good. JavaDoc on the variables would be enough.

File: src/codeu/chat/client/ClientMessage.java [hunk 16]
```diff
@@ -25,6 +25,7 @@
 import codeu.chat.util.Logger;
 import codeu.chat.util.Method;
 import codeu.chat.util.Uuid;
+import javafx.application.Platform;
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientMessage.java:28]: Do you actually use this import somewhere?

File: src/codeu/chat/client/ClientUser.java [hunk 17]
```diff
@@ -18,6 +18,7 @@
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientUser.java:21]: Do you actually use this import anywhere?

File: src/codeu/chat/client/ClientUser.java [hunk 18]
```diff
@@ -123,7 +131,7 @@ public boolean addUser(String name, String password) {
   }
 
   public void showAllUsers() {
-    updateUsers();
+    //updateUsers();
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientUser.java:134]: Why is this commented out? Should it be deleted? Has the functionality been moved?

File: src/codeu/chat/client/ClientUser.java [hunk 19]
```diff
@@ -148,12 +156,23 @@ public String getName(Uuid id) {
   }
 
   public void updateUsers() {
-    usersById.clear();
-    usersByName = new Store<>(String.CASE_INSENSITIVE_ORDER);
-    for (final User user : view.getUsersExcluding(EMPTY)) {
-      usersById.put(user.id, user);
-      usersByName.insert(user.name, user);
-    }
+	if (usersById.size() == 0) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientUser.java:159]: Wait, you're only clearing `usersById` if it's already empty?

File: src/codeu/chat/client/ClientUser.java [hunk 20]
```diff
@@ -148,12 +156,23 @@ public String getName(Uuid id) {
...
+	    usersById.clear();
+	    usersByName = new Store<>(String.CASE_INSENSITIVE_ORDER);
+	    for (final User user : view.getUsersExcluding(EMPTY)) {
+	      usersById.put(user.id, user);
+	      usersByName.insert(user.name, user);
+	    }
+	}
+  }
+  
+  public void updatedUsers(User newUser) {
```
Comment (KevinWorkman) [inline src/codeu/chat/client/ClientUser.java:169]: JavaDoc explaining what this function does, when it should be called, etc.

File: src/codeu/chat/client/Controller.java [hunk 21]
```diff
@@ -33,26 +30,26 @@
 
   private final static Logger.Log LOG = Logger.newLog(Controller.class);
 
-  private final ConnectionSource source;
+  private final Connection source;
```
Comment (KevinWorkman) [inline src/codeu/chat/client/Controller.java:33]: You're probably sick of reading this, but I'd add comments to class-level variables like this. JavaDoc is your friend.

**Review by KevinWorkman — COMMENTED** [review]
Comments comments comments.

I'd also love to see a test for this new stuff. Not totally sure the best way to do that though. Might spend some time thinking about it.

At the very least, you should all hammer away at this to make sure it really works. Try corner cases, one client, many clients, etc. Try to break it.

File: src/codeu/chat/client/View.java (hunk 9 above)
Comment (MalikG15) [reply src/codeu/chat/client/View.java:92]: I don't understand what you mean
